internal/logging/logger.go

```diff
@@ -73,125 +74,145 @@ var (
-		l.updateLogLevel()
+	if level, ok := l.scopeLevels[scope]; ok {
+		return level
 	}
 
-	var scopeLevel zerolog.Level
-	if l.defaultLogLevelFromConfig != -2 {
-		scopeLevel = l.defaultLogLevelFromConfig
-	}
-	if l.defaultLogLevelFromEnv != -2 {
-		scopeLevel = l.defaultLogLevelFromEnv
+	// if the scope is not in the map, use the default level from the root logger
+	if l.defaultLogLevelFromConfig != unset {
+		return l.defaultLogLevelFromConfig
 	}
 
-	// if the scope is not in the map, use the default level from the root logger
-	if level, ok := l.scopeLevels[scope]; ok {
-		scopeLevel = level
+	if l.defaultLogLevelFromEnv != unset {
+		return l.defaultLogLevelFromEnv
 	}
 
-	return scopeLevel
+	return l.defaultLogLevel
 }
 
-func (l *Logger) newScopeLogger(scope string) zerolog.Logger {
-	scopeLevel := l.getScopeLoggerLevel(scope)
-	logger := l.l.Level(scopeLevel).With().Str("component", scope).Logger()
+func (l *Logger) getLogger(scope string) *zerolog.Logger {
+	if logger, ok := l.scopeLoggers[scope]; ok && logger != nil {
+		return logger
+	}
 
-	return logger
-}
+	l.loggerMutex.Lock()
+	defer l.loggerMutex.Unlock()
 
-func (l *Logger) getLogger(scope string) *zerolog.Logger {
-	logger, ok := l.scopeLoggers[scope]
-	if !ok || logger == nil {
-		scopeLogger := l.newScopeLogger(scope)
-		l.scopeLoggers[scope] = &scopeLogger
+	// double-check after acquiring the lock
+	if logger, ok := l.scopeLoggers[scope]; ok && logger != nil {
+		return logger
 	}
 
-	return l.scopeLoggers[scope]
+	scopeLevel := l.getScopeLoggerLevel(scope)
+	logger := l.defaultLogger.Level(scopeLevel).With().Str("component", scope).Logger()
+	l.scopeLoggers[scope] = &logger
+	return &logger
 }
 
-func (l *Logger) UpdateLogLevel(configDefaultLogLevel string) {
-	needUpdate := false
+func (l *Logger) UpdateConfigLogLevel(configDefaultLogLevel string) {
+	var newConfigLevel zerolog.Level
+
+	configDefaultLogLevel = strings.ToUpper(configDefaultLogLevel)
+	loggingContext := l.defaultLogger.With().Str("configDefaultLogLevel", configDefaultLogLevel)
+	logger := loggingContext.Logger()
 
 	if configDefaultLogLevel != "" {
 		if logLevel, ok := zerologLevels[configDefaultLogLevel]; ok {
-			l.defaultLogLevelFromConfig = logLevel
+			logger.Debug().Msgf("setting config log level to %v", logLevel)
+			newConfigLevel = logLevel
 		} else {
-			l.l.Warn().Str("logLevel", configDefaultLogLevel).Msg("invalid defaultLogLevel from config, using ERROR")
-		}
-
-		if l.defaultLogLevelFromConfig != l.defaultLogLevel {
-			needUpdate = true
+			logger.Warn().Msg("invalid log level from config")
+			return
 		}
+	} else {
+		newConfigLevel = unset
 	}
 
-	l.updateLogLevel()
+	l.loggerMutex.Lock()
+	defer l.loggerMutex.Unlock()
 
-	if needUpdate {
-		for scope, logger := range l.scopeLoggers {
-			currentLevel := logger.GetLevel()
-			targetLevel := l.getScopeLoggerLevel(scope)
-			if currentLevel != targetLevel {
-				*logger = l.newScopeLogger(scope)
-			}
+	l.updateLogLevels(newConfigLevel)
```

Comment:
When you change the config-driven log level (not just at startup!) we can now destroy our cache of any affected logger so we recreate a new logger when asked again ... with the correct level.